contracts/axelar-gas-service/src/contract.rs

```diff
@@ -0,0 +1,120 @@
+use soroban_sdk::{
+    contract, contractimpl, panic_with_error, token, Address, Bytes, Env, String, U256,
+};
+
+use axelar_soroban_std::types::{Hash, TokenDetails};
+
+use crate::storage_types::DataKey;
+use crate::{error::Error, event};
+use axelar_soroban_interfaces::axelar_gas_service::AxelarGasServiceInterface;
+
+#[contract]
+pub struct AxelarGasService;
+
+#[contractimpl]
+impl AxelarGasService {
+    pub fn initialize(env: Env, gas_collector: Address) {
+        if env
+            .storage()
+            .instance()
+            .get(&DataKey::Initialized)
+            .unwrap_or(false)
+        {
+            panic!("Already initialized");
+        }
+
+        env.storage().instance().set(&DataKey::Initialized, &true);
+
+        env.storage()
+            .instance()
+            .set(&DataKey::GasCollector, &gas_collector);
+    }
+}
+
+#[contractimpl]
+impl AxelarGasServiceInterface for AxelarGasService {
+    fn pay_gas_for_contract_call(
+        env: Env,
+        sender: Address,
+        destination_chain: String,
+        destination_address: String,
+        payload: Bytes,
+        refund_address: Address,
+        token_details: TokenDetails,
+    ) {
+        sender.require_auth();
+
+        let TokenDetails { token_addr, amount } = token_details.clone();
+
+        if amount == 0 {
+            panic_with_error!(env, Error::InvalidAmount);
+        }
+
+        token::Client::new(&env, &token_addr).transfer(
```

Comment:
[transfer_from](https://docs.rs/soroban-sdk/latest/soroban_sdk/token/struct.TokenClient.html#method.transfer_from)?

contract can't transfer from sender. The tests should avoiding mocking the auth to be able to catch these issues, but need to check how issue a tx from a user address to set it up correctly.

Reply:
in principle i agree with you, but seems like the mechanism is different here, (though i'd like to confirm with them).

* their docs (https://developers.stellar.org/docs/learn/migrate/evm/solidity-and-rust-advanced-concepts#soroban-5) emphasize how the equivalent Solidity convention of requiring approval is replaced/replaceable entirely by the built-in functions in Soroban, e.g. `require_auth`. 
* on the token client implementations (https://github.com/stellar/soroban-examples/blob/v20.0.0/token/src/contract.rs#L100-L127) seem more generalized for both `transfer` and `transfer_from`, in each case accepting both a sender address parameter and receiver address parameter, and there's an auth check on the sender and spender, respectively.
* many of their examples (here, for example: https://github.com/stellar/soroban-examples/blob/v20.0.0/atomic_swap/src/lib.rs#L60), assume the implementation above and have contracts that invoke the `transfer` method when sending funds from a sender to the contract

that said, i'm not entirely sure what the use case is for a `transfer_from` method, as the implementations there seem to invoke transfer from a to b with an allowance from some third party c (where the contract is none of a, b, or c).

Reply:
`account.require_auth()` in it's simplest form means that `account` must be the `msg.sender`. If you compare to an ERC20 example, it's the user/sender who calls ERC20.approve first (in a separate tx). Then the contract (spender), calls transfer_from within the contract. The token contract `transfer_from` checks if the spender is the caller of the method, whereas `transfer` checks that `sender` is the caller. If the contract calls `transfer`, the contract address is the caller, so `sender.require_auth()` will fail. The implementation should translate directly from EVM, take a look at `payGasForContractCall` usage.